### PR TITLE
Store namespaces respective to directories

### PR DESCRIPTION
**Pull request by DanielNoord**
- [x] Add yourself to CONTRIBUTORS if you are a new contributor.
- [x] Write a good description on what the PR does.

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :hammer: Refactoring   |

## Description

This is a first step towards `per-directory-config` as I think they make the most sense:
1) Create a basic `Namespace`
2) Iterate over all `files_or_modules` and see if we can create additional `Namespaces` for subdirectories
3) Look up the correct `Namespace` before checking a file

This is a basic implementation of 3 which allows storing the information that should be gathered in step 2.

### CODE REVIEW

File: pylint/lint/utils.py [hunk 1]
```diff
@@ -99,3 +99,16 @@ def fix_import_path(args: Sequence[str]) -> Iterator[None]:
         yield
     finally:
         sys.path[:] = original
+
+
+def _is_relative_to(self: Path, *other: Path) -> bool:
```
Comment (Pierre-Sassoulas) [inline pylint/lint/utils.py:104]:
```suggestion
def pathlib_relative_to_is_callable(self: Path, *other: Path) -> bool:
```
? I don't think I understand what this function does. Shouldn't the value be returned too ?

File: tests/config/test_per_directory_config.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    runner = Run([__name__], exit=False)
+    assert id(runner.linter.config) == id(runner.linter._base_config)
+
+    # When the file is a directory that does not have any of its parents in
+    # linter._directory_namespaces it should default to the base config
+    with tempfile.TemporaryDirectory() as tmpdir:
+        with open(Path(tmpdir) / "test.py", "w", encoding="utf-8") as f:
+            f.write("1")
+        Run([str(Path(tmpdir) / "test.py")], exit=False)
+        assert id(runner.linter.config) == id(runner.linter._base_config)
```
Comment (Pierre-Sassoulas) [inline tests/config/test_per_directory_config.py:24]:
```suggestion
    with open(test_file, "w", encoding="utf-8") as f:
        f.write("1")
    Run([str(test_file)], exit=False)
    assert id(runner.linter.config) == id(runner.linter._base_config)
```

File: tests/config/test_per_directory_config.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+# For details: https://github.com/PyCQA/pylint/blob/main/LICENSE
+# Copyright (c) https://github.com/PyCQA/pylint/blob/main/CONTRIBUTORS.txt
+
+import tempfile
+from pathlib import Path
+
+from pylint.lint import Run
+
+
+def test_fall_back_on_base_config() -> None:
```
Comment (Pierre-Sassoulas) [inline tests/config/test_per_directory_config.py:11]:
```suggestion
def test_fall_back_on_base_config(tmpdir: LocalPath) -> None:
```

There's a pytest fixture for that.

File: tests/config/test_per_directory_config.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+    """Test that we correctly fall back on the base config."""
+    # A file under the current dir should fall back to the highest level
+    # For pylint this is ./pylintrc
```
Comment (Pierre-Sassoulas) [inline tests/config/test_per_directory_config.py:14]:
```suggestion
    # For pylint this is ./pylintrc
     test_file = tmpdir /  "test.py"
```

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]

File: pylint/lint/utils.py (hunk 1 above)
Comment (DanielNoord) [reply pylint/lint/utils.py:104]: It's just a line for line copy of the `stdlib` function. We don't really need to return anything else?

File: tests/config/test_per_directory_config.py [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+from pathlib import Path
+
+from pylint.lint import Run
+
+
+def test_fall_back_on_base_config(tmpdir: LocalPath) -> None:
+    """Test that we correctly fall back on the base config."""
+    # A file under the current dir should fall back to the highest level
+    # For pylint this is ./pylintrc
+     test_file = tmpdir /  "test.py"
```
Comment (DanielNoord) [inline tests/config/test_per_directory_config.py:15]:
```suggestion
    test_file = tmpdir /  "test.py"
```

**Review by Pierre-Sassoulas — APPROVED** [review]

**Pull request merged**